### PR TITLE
Add passthrough success and failure result methods

### PR DESCRIPTION
**Pull request by seanim0920**
There are a few scenarios where we need to perform side effects after a success result, and mark the whole operation as a failure if these side effects fail.

Ex.
```ts
createEvent(tx, req.body, res.locals.selfId)
  .flatMapSuccess(({ insertId }) =>
    addHostToAttendeeList(
      tx,
      res.locals.selfId,
      parseInt(insertId),
      "hosting"
    ).withSuccess(({ insertId }))
  )
```

In this example we have two operations, createEvent and addHostToAttendeeList, which is dependent on createEvent succeeding. If addHostToAttendeeList succeeds, we still want to return the success value of the initial createEvent operation. If addHostToAttendeeList fails, we want to return the failure of addHostToAttendeeList to provide a clear reason for failure. To accomplish this with the current operations we'd need to combine a flatMapSuccess and a withSuccess. This PR adds an operation "passthroughSuccess" to reduce boilerplate.

```ts
createEvent(tx, req.body, res.locals.selfId)
  .passthroughSuccess(({ insertId }) =>
    addUserToAttendeeList(
      tx,
      res.locals.selfId,
      parseInt(insertId),
      "hosting"
    )
  )
```

For completeness I've also added a "passthroughFailure" method, which will pass the current failure value through if the next result is successful, or if it's not, override the current failure. Haven't seen a purpose for this yet though.

### CODE REVIEW

File: lib/Result.ts [hunk 1]
```diff
@@ -60,6 +60,28 @@ export class SuccessResult<Success, Failure> {
...
 
+  /**
+   * Runs the handler with the current success value and passes the current success value through
+   * if the handler is successful.
+   *
+   * @param handler a function to process the current success value and either passthrough the value or return a failure.
+   */
+  passthroughSuccess<NewSuccess, NewFailure>(
+    handler: (value: Success) => AwaitableResult<NewSuccess, NewFailure>
+  ): SuccessResult<Success, NewFailure | Failure> {
```
Comment (mhayes853) [inline lib/Result.ts:71]:
```suggestion
  ): AwaitableResult<Success, NewFailure | Failure> {
```
The handler may return a failure or promise result.

File: lib/Result.ts [hunk 2]
```diff
@@ -161,6 +183,28 @@ export class FailureResult<Success, Failure> {
...
+  
+  /**
+   * Runs the handler with the current failure value and passes the current failure value through
+   * if the handler is successful.
+   *
+   * @param handler a function to process the current failure value and either passthrough the value or return a failure.
+   */
+  passthroughFailure<NewSuccess, NewFailure>(
+    handler: (value: Failure) => AwaitableResult<NewSuccess, NewFailure>
+  ): FailureResult<Success, NewFailure | Failure> {
```
Comment (mhayes853) [inline lib/Result.ts:205]:
```suggestion
  ): AwaitableResult<Success, NewFailure | Failure> {
```
See same comment on `SuccessResult`.

File: lib/Result.ts [hunk 3]
```diff
@@ -161,6 +183,28 @@ export class FailureResult<Success, Failure> {
...
+    return handler(this.value).flatMapSuccess(() => failure(this.value))
```
Comment (mhayes853) [inline lib/Result.ts:206]:
```suggestion
    return handler(this.value).withFailure(this.value)
```

File: lib/Result.ts [hunk 4]
```diff
@@ -22,7 +22,7 @@ export type AwaitableResult<Success, Failure> =
 /**
  * A result representing the success of an operation.
  */
-export class SuccessResult<Success, Failure> {
+class SuccessResult<Success, Failure> {
```
[the remark this answers is not in the document]
Comment (mhayes853) [reply lib/Result.ts:25]: If we don't export them, it's harder to add extensions or do `instanceof` checks if necessary.

[the remark this answers is not in the document]
Comment (seanim0920) [reply lib/Result.ts]: fixed

File: lib/Result.ts (hunk 1 above)
Comment (seanim0920) [reply lib/Result.ts]: fixed

File: lib/Result.ts (hunk 2 above)
Comment (seanim0920) [reply lib/Result.ts]: fixed

File: lib/Result.ts (hunk 3 above)
Comment (seanim0920) [reply lib/Result.ts:207]: no need for this, this method is actually for converting the success into a failure.

**Review by mhayes853 — APPROVED** [review]

**Pull request merged**